Future/babanuki

ddtTamari opened the pull request.

ddtTamari committed:
課題番号work6-6の修正版

当日行った内容  100％
・今後の管理がしやすいようにクラスを追加
・上記に伴いシーケンス図を修正

ddtTamari committed:
[WIP]課題番号 なし

当日行った内容  ５％
次回todo
・カードクラスの使用済みカードの格納メソッド
・進行役クラスのカード配布処理

ddtTamari committed:
[WIP]課題番号 ババ抜き

当日行った内容  13％
次回todo
・Playerクラスのコンストラクタの作成
・メインメソッドでPlayerクラスの生成

ddtTamari committed:
[WIP]課題番号　なし

当日行った内容  15％
次回todo
・手札クラスの作成
・カードクラスの処理の作成

ddtTamari committed:
[WIP]課題番号ババ抜き

当日行った内容  20％
次回todo
・カードを手札に渡す処理
・進行役の動作

ddtTamari committed:
[WIP]課題番号 ババ抜き

当日行った内容  22％
次回todo
・カードをプレイヤーに配る

ddtTamari committed:
[WIP]課題番号 ババ抜き
・カードを手札に配る
当日行った内容  30％
次回todo
・そろったカードを捨てる
・最後のプレイヤーにカードが配られていない不具合を修正

ddtTamari committed:
[WIP]課題番号 ババ抜き

当日行った内容  35％
次回todo
・そろっていない数字が捨てられてしまう不具合の修正
・手札の表示が数字になっているところをトランプ名に合わせるよう変更するメソッドの追加

ddtTamari committed:
[WIP]課題番号 ババ抜き

当日行った内容  31％
次回todo
・そろっていないカードが捨てられる不具合を修正

ddtTamari committed:
[WIP]課題番号ババ抜き

当日行った内容   カードマネージャークラスの作成　38％
次回todo
・そろったカードを捨てる処理

ddtTamari committed:
[WIP]課題番号 ババ抜き

当日行った内容 手札が配られた後にそろったカードを捨てる処理 40％
次回todo
・隣の人からカードを引く処理

ddtTamari committed:
[WIP]課題番号ババ抜き

当日行った内容  41％
・プレイヤー人数が少人数の場合にすべての手札の組み合わせで同じ数字かどうかの確認ができていなかった不具合の修正
次回todo
・他人からカードを引く処理の検討

ddtTamari committed:
[WIP]課題番号babanuki

当日行った内容  43％
・複数クラスでの共通処理を共通化
・捨てカードリストの登録と表示
次回todo
・ほかプレイヤーからカードを引く

ddtTamari committed:
[WIP]課題番号 ババ抜き

当日行った内容  47％
・進行役とプレイヤークラスがプレイヤーからカードを引く処理を追加
次回todo
・プレイヤーからカードを引く処理の完成

ddtTamari committed:
[WIP]課題番号ババ抜き

当日行った内容  50％
・つぎの人からカードを引く処理
次回todo
・一番最後の人がカードを引けない不具合を修正
・カードを引いた後そろっていたなら捨てる処理を追加

ddtTamari committed:
[WIP]課題番号 ババ抜き

当日行った内容  60％
・最後の人が隣の人のカードを引けない不具合の修正
・引いたカードがそろった時に捨てる処理を追加
・プレイヤーのリストを作成
次回todo
・カードが0枚になったプレイヤーをプレイヤーリストから削除する

ddtTamari committed:
[WIP(途中の場合のみ付ける)]課題番号babanuki

当日行った内容+80％(全体の割合)
・手札確認を終わった人を抜かす
次回todo
・人数が減った時のカードを引く動作の確認
・終了処理

ddtTamari committed:
[WIP]課題番号 babanuki

当日行った内容  85％
・最後の一人になったときに順位を宣言する
・プレイヤーの間の人が上がったときにその次の人の引く処理がスキップされる不具合を修正
次回todo
・途中で真ん中の人が上がったときに落ちる不具合を修正
・コメントを追加
・今まで行っていたデバッグ用処理を削除

ddtTamari committed:
[WIP]課題番号 babanuki

当日行った内容  88％
真ん中の人のカードを引いたときにエラーが起きる不具合を修正
次回todo
・最後の人の一人前の人が上がったときカードを引く人が正しくない不具合を修正
・コメントを付ける

ddtTamari committed:
[WIP]課題番号babanuki

当日行った内容 89％
・必要のない引数の削除
・不具合内容の確認
次回todo
・要素数がオーバーする不具合の修正

ddtTamari committed:
[WIP]課題番号babanuki

当日行った内容  98％
・配列要素数を超える不具合の修正
・コメント付け
次回todo
・コメント付け

ddtTamari committed:
[WIP]課題番号babanuki

当日行った内容  95％
コメント付け
不必要引数の削除
次回todo
Javadocつけ

ddtTamari committed:
課題番号 babanuki

当日行った内容  100％
・Javadocを付ける

ddtTamari committed:
課題番号 babanuki

当日行った内容  100％
・メンバ変数を同一クラス内で引数に渡しているところを修正しました。

ddt-mentor commented on babanuki/Constant.java:
```diff
@@ -0,0 +1,57 @@
+    public static final int MISSING_NUM_111110 = 62; //ビット管理時に使用しない数字
+    public static final int MISSING_NUM_ELEMENT = 7; //ビット管理時に使用しない数字
+
+    //出力メッセージ
```
海外展開の観点から、システムメッセージに関する定数は別ファイルに切り出すと良いです。
(翻訳担当者に渡す際に便利なので)
by石井

ddt-mentor commented on babanuki/Constant.java:
```diff
@@ -0,0 +1,57 @@
+    public static final String FINISH_PLAYER_MASSAGE = "さんは上がりました。";
+    public static final String RANK_MESSAGE = "位は";
+    public static final String PLAYER_MESSAGE = "さんです";
+    public static final String DROP_CARD_MASSAGE = "は捨てられた…";
```
DROP_CARD_MESSAGE ?
石井

ddt-mentor commented on babanuki/Constant.java:
```diff
@@ -0,0 +1,57 @@
+
+public class Constant {
+    //共通用定数
+    public static final int INITIAL_NUM = 0; //値を初期化する汎用定数
```
後でチェックしておく(メモ)

ddt-mentor replied:
INITIAL_NUMが、カード、プレイヤー、for文のiといった複数の対象に対する初期値として用いられています。
カードとプレイヤーはそれぞれ定数を分け、for文の場合は単に0とする方が、可読性や可用性の観点から良いと思います。(石井)

ddt-mentor commented on babanuki/ConvertTrumpCard.java:
```diff
@@ -0,0 +1,59 @@
+        //スート用変数
+        String suit = "";
+        //jokerだけは別枠で強制的にjokerマークにする
+        if (card == Constant.TRUMP_SUIT_JOKER) {
```
規約確認しておく(メモ)

ddt-mentor commented on babanuki/HandsManage.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    /**
+     * メソッド名：setPlayerHand
```
メソッド名が違います(石井)

ddt-mentor commented on babanuki/HandsManage.java:
```diff
@@ -0,0 +1,62 @@
+     * 処理内容：カードをプレイヤーの手札に登録する
+     */
+    //プレイヤーの手札に登録する
+    public void setPlayerHand(int card) {
```
setPlayerHandとgetPlayerHandに対称性がないです。
この場合は手札を加えるわけですから、addPlayerHandが良いのではないでしょうか(石井)

ddt-mentor commented on babanuki/Facilitator.java:
```diff
@@ -0,0 +1,285 @@
+     */
+    // ゲーム開始処理
+    public void doOldMaid() {
+        int playerNum = Constant.INITIAL_NUM; //  プレイヤーの人数を格納する変数
```
この行は必要ないです。
最初からaskPlayerNum()を行うことで、代入コストが減ります。(石井)

ddt-mentor commented on babanuki/Player.java:
```diff
@@ -0,0 +1,188 @@
+        //引いたプレイヤーの手札に新たに引いたカードを追加する
+        setPlayerHand(opponent.retHands(drawPosition));
+        //引かれたプレイヤーのカードを削除する
+        opponent.removePlayerHand(drawPosition);
```
プレイヤーが他のプレイヤーの値を操作する形はババ抜きの絵面としては自然ですが、オブジェクトの役割としては不自然です。
この場合は手札を引数として受け取って、その中からカードを引くという形が望ましいのではないかと思います。
（石井）

ddt-mentor commented:
メンバ変数には先頭にｍをつけましょう

ddt-mentor commented on babanuki/BabanukiManage.java:
```diff
@@ -0,0 +1,88 @@
+        //もともとの手札の数字だけで見るための変数
+        int handsCardNum = Constant.INITIAL_NUM;
+        //同じ数字かどうか
+        boolean sameNum = false;
```
booleanには命名規則に割りと目立ったルールがあります。
メソッドでも変数でもis○○やhas○○という書き方をすることが多いです。（他にもenable等）
もちろん他にも命名規則として細かい部分はありますが、これは割りと他との違いが顕著なルール
なので、確認後反映をしてみてください。
佐藤

ddt-mentor commented on babanuki/BabanukiManage.java:
```diff
@@ -0,0 +1,88 @@
+     * 処理内容：指定された要素番号のカードの値を返す
+     */
+    //指定されたindexの数字を返す
+    public int retCard(int getCardId) {
```
retは命名に使わないほうがいいです。どこまでNGなのかは難しい部分ですが、基本的に省略系は避けるようにしましょう
佐藤

ddt-mentor replied:
もうひとつ。仮引数名にメソッドの命名規則が使われています。表現するのが難しいという話はあるかもしれないですが、違う名前を考えてみてください。
佐藤

ddt-mentor commented on babanuki/BabanukiManage.java:
```diff
@@ -0,0 +1,88 @@
+     * @return
+     * 処理内容：カードが配られたときに手札を確認するメソッド
+     */
+    //最初に同じ番号があるかどうかを確認するメソッド
```
javadocのこの部分のコメントは、メソッド名とここで書いてある下に書きましょう。確認してないですが、javadocを作るときに不備が出る恐れがあります。

ddt-mentor commented on babanuki/DiscardTable.java:
```diff
@@ -0,0 +1,49 @@
+     * @param  discardCard : カードの値
+     *          discardCardTwo：カードの値2
+     * @return
+     * 処理内容：捨てられたカードを表示する
```
ここに書くと、returnの値としてjavadocで判断されてしまう可能性があります。
佐藤

ddt-mentor commented on babanuki/Facilitator.java:
```diff
@@ -0,0 +1,285 @@
+
+public class Facilitator {
+    // キーボードからの入力を取得する
+    Scanner userImput = new Scanner(System.in);
```
これはuserInputのtypoに見えます。ご確認お願いします。
佐藤

ddtTamari committed:
課題番号 babanuki修正版①

当日行った内容  100％
・システムメッセージを"MessageConstant"クラスに分けました。
・Javadocのメソッド名を本当のメソッド名に修正しました。
・”setPlayerHand”メソッドを”addPlayerHand”に修正しました。
・不必要な代入処理(0での初期化)を削除しました。
・INITIAL_NUMを”PLAYER_INITIAL_ NUM”と”CARD_INITIAL_NUM”に変更し、For文に使用されていたINITIAL_NUMを0に変更しました。
・プレイヤーが次のプレイヤーから手札を引く動作がプレイヤークラスを渡して行っていたところを手札を渡して引くよう修正しました。
・メンバ変数に”m”を付けました。
・Booleanの変数/メソッドを命名規則に沿うよう修正しました。

ddt-mentor commented:
@ddtTamari 
山本　補足
1　ババ抜きを行う際に、「最初の順番」が必ずプレイヤー1から始まるようになっています。
　　ただ、テキスト（仕様）には記載されていない内容なので、対応は必須とはしません。
　　対応の例としては、ランダム数値を生成して、順番を決めたりジャンケンゲームを前の章でやっているのでそれを用いたりする、という事例があります。

2　定数クラスについて、カードのスートなどの定義は他のゲームでも使えるので、
　　カード用の定数クラスは、外出しにする、というのもありです。（パッケージを分ける）
　　ただし、同一パッケージの統一感がなくなる場合もあるので、一概に外出しがいいとは限らない。

ddt-mentor commented on babanuki/Player.java:
```diff
@@ -0,0 +1,200 @@
+    //コンストラクタ群
+
+    public Player() {
+    }
```
柴田(補足)
Player作成時、名前が与えられなかった場合のコンストラクタと読み取りますが、使用していないようです。
今回のプログラムの場合、名前がなくても使用箇所がないため特段問題にはなりませんが、
実際は使えないようにしておくべきと考えられます。基本は削除します。
(聞いた話レベルですが、明示的に引数無しが使用できないことを示すため、privateメソッドにすることもあるようです)

ddt-mentor commented on babanuki/Facilitator.java:
```diff
@@ -0,0 +1,288 @@
+
+        String[] userName = new String[playerNum]; // ユーザの名前を格納しておく配列
+        Player[] player = new Player[playerNum]; // 人数分のプレイヤーを格納する配列
+        List<Player> playerList = new ArrayList<Player>(); // 現在プレイしているプレイヤーリスト
```
柴田
少し細かいのですが、UserNameはPlayerが持つべきではないでしょうか。
使用されていませんが、PlayerクラスにはmUserNameが存在しています。
少なくとも、ユーザーと名前が別のオブジェクトで管理されてしまっていることは違和感があります。
(影響度合いが大きく修正に時間がかかるようであれば、今すぐ直さなくて良いです)
